coderd/database/dbmem/dbmem.go

```diff
@@ -1204,12 +1265,74 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppSecretByID(_ context.Context, id uu
 		if secret.ID == id {
 			q.oauth2ProviderAppSecrets[index] = q.oauth2ProviderAppSecrets[len(q.oauth2ProviderAppSecrets)-1]
 			q.oauth2ProviderAppSecrets = q.oauth2ProviderAppSecrets[:len(q.oauth2ProviderAppSecrets)-1]
+
+			// Cascade delete into tokens.
+			var keyIDsToDelete []string
+			var tokens []database.OAuth2ProviderAppToken
+			for _, token := range q.oauth2ProviderAppTokens {
+				if token.AppSecretID == id {
+					keyIDsToDelete = append(keyIDsToDelete, token.APIKeyID)
+				} else {
+					tokens = append(tokens, token)
+				}
+			}
+			q.oauth2ProviderAppTokens = tokens
+
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
+			q.apiKeys = keys
+
 			return nil
```

Comment:
nesting all this logic in the for loop feels a bit difficult to read. Maybe it's just me, but this feels a bit cleaner imo.

```golang
	index := slices.IndexFunc(q.oauth2ProviderAppSecrets, func(secret database.OAuth2ProviderAppSecret) bool {
		return secret.ID == id
	})

	if index < 0 {
		return sql.ErrNoRows
	}

	q.oauth2ProviderAppSecrets[index] = q.oauth2ProviderAppSecrets[len(q.oauth2ProviderAppSecrets)-1]
	q.oauth2ProviderAppSecrets = q.oauth2ProviderAppSecrets[:len(q.oauth2ProviderAppSecrets)-1]
	// ... rest of delete ...

```

Reply:
You are right, this is much better.

Reply:
I do this in a few places, going to update all of them along with using `DeleteFunc`!